evidence/verify.go

```diff
@@ -48,62 +52,59 @@ func (evpool *Pool) verify(evidence types.Evidence) (*info, error) {
 	case *types.DuplicateVoteEvidence:
 		valSet, err := evpool.stateDB.LoadValidators(evidence.Height())
 		if err != nil {
-			return nil, err
+			return err
 		}
-		err = VerifyDuplicateVote(ev, state.ChainID, valSet)
-		if err != nil {
-			return nil, fmt.Errorf("verifying duplicate vote evidence: %w", err)
-		}
-
-		_, val := valSet.GetByAddress(ev.VoteA.ValidatorAddress)
-
-		return &info{
-			Evidence:         evidence,
-			Time:             evTime,
-			Validators:       []*types.Validator{val}, // just a single validator for duplicate vote evidence
-			TotalVotingPower: valSet.TotalVotingPower(),
-		}, nil
+		return VerifyDuplicateVote(ev, state.ChainID, valSet)
 
 	case *types.LightClientAttackEvidence:
 		commonHeader, err := getSignedHeader(evpool.blockStore, evidence.Height())
 		if err != nil {
-			return nil, err
+			return err
 		}
 		commonVals, err := evpool.stateDB.LoadValidators(evidence.Height())
 		if err != nil {
-			return nil, err
+			return err
 		}
 		trustedHeader := commonHeader
 		// in the case of lunatic the trusted header is different to the common header
 		if evidence.Height() != ev.ConflictingBlock.Height {
 			trustedHeader, err = getSignedHeader(evpool.blockStore, ev.ConflictingBlock.Height)
 			if err != nil {
-				return nil, err
+				return err
 			}
 		}
 
 		err = VerifyLightClientAttack(ev, commonHeader, trustedHeader, commonVals, state.LastBlockTime,
 			state.ConsensusParams.Evidence.MaxAgeDuration)
 		if err != nil {
-			return nil, err
+			return err
 		}
 		// find out what type of attack this was and thus extract the malicious validators. Note in the case of an
 		// Amnesia attack we don't have any malicious validators.
-		validators, attackType := getMaliciousValidators(ev, commonVals, trustedHeader)
-		totalVotingPower := ev.ConflictingBlock.ValidatorSet.TotalVotingPower()
-		if attackType == lunaticType {
-			totalVotingPower = commonVals.TotalVotingPower()
+		validators := ev.GetByzantineValidators(commonVals, trustedHeader)
+		// ensure this matches the validators that are listed in the evidence. They should be in the same order as
+		// that of the commit.
+		if len(validators) != len(ev.ByzantineValidators) {
+			return fmt.Errorf("expected %d byzantine validators from evidence but got %d",
+				len(validators), len(ev.ByzantineValidators))
 		}
 
-		return &info{
-			Evidence:         evidence,
-			Time:             evTime,
-			Validators:       validators,
-			TotalVotingPower: totalVotingPower,
-		}, nil
+		for idx, val := range validators {
+			if !bytes.Equal(ev.ByzantineValidators[idx].Address, val.Address) {
```

Comment:
Should we define a method on `Validator`, say `Equal(other Validator)` cause I think I've seen this before

Reply:
Yeah but I also wanted to avoid comparing pubkeys and proposer priority as this information is a bit useless in the case of the abci evidence. Originally I wanted to just have an array of `abci.Validator`'s but then this would cause circular imports